aten/src/ATen/native/native_functions.yaml

```diff
@@ -10169,6 +10162,27 @@
   dispatch:
     Math: linalg_tensorsolve_out
 
+- func: linalg_qr(Tensor self, str mode='reduced') -> (Tensor Q, Tensor R)
+  python_module: linalg
+  use_c10_dispatcher: full
+  variants: function
+  dispatch:
+    DefaultBackend: linalg_qr
+
+- func: linalg_qr.out(Tensor self, str mode='reduced', *, Tensor(a!) Q, Tensor(b!) R) -> (Tensor(a!) Q, Tensor(b!) R)
+  use_c10_dispatcher: hacky_wrapper_for_legacy_signatures
+  python_module: linalg
+  variants: function
+  dispatch:
+    DefaultBackend: linalg_qr_out
```

Comment:
What's the distinction between DefaultBackend and math dispatch here?

In practice we could probably make all these dispatches CPU, CUDA:, since ultimately these functions are just wrappers around _linalg_qr_helper. I'm not sure there's much value in making them Math or DefaultBackend.

Reply:
I'm not sure to understand the question. In this branch `linalg_qr` and `linalg_qr_out` are the "main" functions: they have a corresponding entry in `derivatives.yaml` and thus they require `DefaultBackend`, if I understand correctly how the `dispatch:` mechanism works.
On the other hand `qr` and `qr_out` are implemented in terms of `linalg_qr` and thus they use `dispatch: Math`.

For the record, I tried to use `Math` for `linalg_qr_out` but as expected I get this compilation error:
```
AssertionError: There's a formula for linalg_qr_out(or its functional variant) in derivatives.yaml. It's required to add a dispatch section for it with explicit supported backends e.g CPU/CUDA or DefaultBackend in native_functions.yaml. Please see https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native#choosing-the-right-dispatch-keyword for instructions to choose the right dispatch keyword.
```

Reply:
Interesting, I didn't realize that we were tying "out" variants to derivatives.yaml.

@albanD @smessmer @ailzhang  - we should probably update the developer FAQ (https://github.com/pytorch/pytorch/wiki/Developer-FAQ) with a description of when to use each dispatch and how entries are associated in autograd.yaml with a couple examples.

Reply:
I don't think we should have a formula for any ".out" variant. And the codegen would prevent that I think if you use the naming convention with the `.`